content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,327 @@
+
+In the loading example, there is a clear source of truth for both the "source" prop and the "loading" state. When the source prop changes, the loading state should **always** be overridden. Conversely the state should only be overridden when the prop **changes** and should otherwise be managed by the component.
+
+Problems arise when any of these constraints are changed. This typically comes in two forms. Let's take a look at both.
+
+### Anti-pattern: Props always override state
+
+A common misconception about both `getDerivedStateFromProps` and `componentWillReceiveProps` is that they are only called when props "change". In reality, these lifecycles are called any time a component is re-rendered. (As of 16.4, `getDerivedStateFromProps` is also called for renders caused by `setState` or `forceUpdate`.) Because of this, it is unsafe to unconditionally override state values using either of these lifecycles.
+
+Here is a `EmailInput` component that accepts an email prop:
+```js
+class EmailInput extends Component {
+  state = {
+    email: this.props.email
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // This is bad! Do not do this!
+    if (props.email) {
+      return { email: props.email };
+    }
+    return null;
+  }
+
+  render() {
+    return <input onChange={this.handleChange} value={this.state.email} />;
+  }
+
+  handleChange = event => this.setState({ email: event.target.value });
+}
+```
+
+At first, this component might look okay. State is initialized to the value specified by props and updated when we type into the `<input>`. But once our component rerenders– either because it called `setState` or because its parent rerendered– anything we've typed into the `<input>` will be lost!
+
+At this point, you might be wondering if this component would have worked with version 16.3. Unfortunately, the answer is "no". Before moving on, let's take a look at why this is.
+
+For the simple example above, we could "fix" the problem of unexpected rerenders using `shouldComponentUpdate`. However in practice, components usually accept multiple props, and our component would rerender if any one of them changed. Even if none of them "changed", props that are functions or objects are often created inline and so will always bypass `shouldComponentUpdate`. For example, what if our component accepted a function to validate the current email address?
+```js
+<EmailInput
+  email={this.props.user.email}
+  validate={this.validateEmail.bind(this)}
+/>
+```
+
+The above example binds the validation callback inline and so it will pass a new function prop every time it renders- effectively bypassing `shouldComponentUpdate` entirely. [Here is a demo](codesandbox://when-to-use-derived-state/derived-state-anti-pattern) that uses a timer to illustrate this problem. Since you can't prevent others from using your component this way, it's inherently fragile.
```

Comment:
I don't want to get into the "other data" explanation quite yet, because it bleeds into the technique of checking for the specific changed prop, which isn't introduced quite yet.

Reply:
I'll post a re-wording of this section. Does it help?

Reply:
Yeah this is great